### ISSUE TITLE
Deserializing some dates into java.util.Date with timezone CET is wrong

### ISSUE THREAD

**Issue by modrljin** [FasterXML/jackson-databind#3519]
I believe deserialization of some dates doesn't work correctly.

e.g. string '1940-04-02' or:

Date: Tue Apr 02 00:00:00 CET 1940
Date formatted: 1940-04-02
Datetime formatted: 1940-04-02 00:00:00
Datetime formatted with zone: 1940-04-02 00:00:00 +0100

is deserailized into:

@JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd ", timezone = "CET")
**CET date: 1940-04-01**

@JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd ", timezone = "UTC")
UTC date: 1940-04-02

@JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss", timezone = "CET")
**CET datetime: 1940-04-01 23:00:00 +0100**

@JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss", timezone = "UTC")
UTC datetime: 1940-04-02 01:00:00 +0100

I have made additional test to check all the periods from year 1900 until 2100 and there are a lot of dates when this issue happens:

Period: 1916-05-01 - 1916-09-30
Period: 1917-04-17 - 1917-09-17
Period: 1918-04-16 - 1918-09-16
Period: 1940-04-02 - 1941-04-18
Period: 1945-04-03 - 1945-05-08
Period: 1977-04-04 - 1977-09-25
Period: 1978-04-03 - 1978-10-01
Period: 1979-04-02 - 1979-09-30
Period: 1980-04-07 - 1980-09-28
Period: 1981-03-30 - 1981-09-27
Period: 1982-03-29 - 1982-09-26

Other days are correctly deserialized.

**Comment by cowtowncoder** [comment]
There may be a misunderstanding here wrt role of `timezone` parameter for `@JsonFormat` annotation: it only affects serialization and specifically does not affect deserialization (`java.util.Date` does not have timezone information; it is a strict timestamp).
Maybe that helps explain what you seeing.

In addition, textual description is not really enough: a code reproduction (unit test) would be needed, to show exact failure(s). It would be enough to start with a single case.

**Comment by modrljin** [comment]
I'm aware that java.util.Date doesn't have timezone information, I'm just printing it in my local timezone CET with SimpleDateFormat. But this doesn't mean that it shouldn't be correctly deserialized into java.util.Date considering timezone parameter. It works fine for other dates which are outside of the periods I wrote before, e.g.

Date: Mon Mar 29 00:00:00 CET 1943 (timestamp: -844563600000)
Date formatted: 1943-03-29
Datetime formatted: 1943-03-29 00:00:00
Datetime formatted with zone: 1943-03-29 00:00:00 +0100
-----------
CET date: 1943-03-29
UTC date: 1943-03-29
CET datetime: 1943-03-29 00:00:00 +0100 (timestamp: -844563600000)
UTC datetime: 1943-03-29 01:00:00 +0100 (timestamp: -844560000000)
###########

Date: Tue Mar 30 00:00:00 CEST 1943 (timestamp: -844480800000)
Date formatted: 1943-03-30
Datetime formatted: 1943-03-30 00:00:00
Datetime formatted with zone: 1943-03-30 00:00:00 +0200
-----------
CET date: 1943-03-30
UTC date: 1943-03-30
CET datetime: 1943-03-30 00:00:00 +0200 (timestamp: -844480800000)
UTC datetime: 1943-03-30 02:00:00 +0200 (timestamp: -844473600000)
###########

Btw. when deserializing with timezone "Europe/Zagreb" all dates between year 1900-2100 are deserialized correctly.
As far I can see, in most cases in all those periods when it doesn't work fine it is a around a start date of CEST (daylight saving time or +0200 instead of CET +0100).

I'll attach existing code I have used for testing, it is without unit test, but maybe that will be enough for now
[TestJson.txt](https://github.com/FasterXML/jackson-databind/files/8914130/TestJson.txt)
[TestObject.txt](https://github.com/FasterXML/jackson-databind/files/8914131/TestObject.txt)
.

**Comment by yawkat** [comment]
I cannot reproduce the issue with the test case you provided. It only prints 'CORRECT' for me.

Also, I can only recommend avoiding java.util.Date in favor of the java.time APIs introduced in java 8. Timezone handling in the old API is particularly messy, there is no reason to keep using it.

**Comment by modrljin** [comment]
Please try to comment line 63: //printDate(date, o, true); so it doesn't print out the correct dates because there is a lot of them and maybe you don't see it in your console. You should see then also INVALID dates. If you get it you can then try to comment line 53: //printDate(date, o, false); as well to get only info about periods where the problem occurs.

**Comment by yawkat** [comment]
with the expanded date range, i can see the 'INVALID' branch, but I still don't see how the particular run that triggers it is wrong. For the run that fails, `date` is `Fri May 25 00:00:00 CEST 1945` (`-776487600000`). Your test case converts this to the string `1945-05-25 00:00:00`, which is correct in my timezone (CEST). Jackson then deserializes this string into the `cetDatetime` field, giving `Fri May 25 01:00:00 CEST 1945`. This is also correct, because `1945-05-25 00:00:00` in CET is the same instant as `1945-05-25 01:00:00` in CEST. So in this case, the test case prints 'INVALID' even though the discrepancy between the getTime of the two values is expected – since when you wrote out the date as a string, you did not include the timezone, you can get a different `getTime` when deserializing.

The test case is very brittle because it depends on system time zone of the person running it. I think the reason why you are failing on different years than I am is that we live in different countries (albeit both roughly in UTC+1). For example, the above failure in May 1945 that I saw happened because the Europe/Berlin timezone (where I live) [changed time zone rules](https://github.com/eggert/tz/blob/0e8f0b06ac2f65d0c54a91121da09c8662cd88e6/europe#L1489-L1493) at that date. This makes the `dateAddDays` of your test case change the timezone of the `date` variable from CET to CEST.

Please make the test case less sensitive to the system time zone. This would help us a lot in reproducing your issue reliably.

**Comment by modrljin** [comment]
When I create Date from timestamp: `-776487600000` I get `Thu May 24 23:00:00 CEST 1945` and you say it is `Fri May 25 00:00:00 CEST 1945` on your system ??? That's really weird. When you check any online time converter tool it converts into my time, i.e. https://www.unixtimestamp.com, https://www.epochconverter.com/, https://timestamp.online/ ...

E.g. on webiste https://www.unixtimestamp.com it is as follows:
`Enter a timestamp: -776487600`
result...
`Your Time Zone: Thu May 24 1945 23:00:00 GMT+0200 (Central European Summer Time)`

It looks like that is the first problem when you run my test :-( I thought that timestamp is really a timestamp in java and creating Date from it doesn't create or print different objects depending of the environment or timezone. Can you check this once again? Try following:

```
String sDate = "1945-05-25";
Date date1 = parseDate(sDate, TestObject.DATE_FORMAT, true, TimeZone.getTimeZone("CET"));
System.out.println("Date1: " + date1 + " (timestamp: " + date1.getTime() + ")");

Date date2 = new Date(-776487600000L);
System.out.println("Date2: " + date2 + " (timestamp: " + date2.getTime() + ")");
```

Result should be:

```
Date1: Fri May 25 00:00:00 CEST 1945 (timestamp: -776484000000)
Date2: Thu May 24 23:00:00 CEST 1945 (timestamp: -776487600000)
```

Once we figure out what's wrong here then I can try to create a better test case. But probably next week since I'm going to a long weekend :-)

And just to be clear about the basic fact, do you agree that deserializing string `1940-04-02 00:00:00` with jackson into `cetDatetime` field should produce Date `Tue Apr 02 00:00:00 CET 1940` and not `Mon Apr 01 23:00:00 CET 1940` as it is when I run it? At least deserialization for all dates after 1983 on my system always produces Date `Day Month DD 00:00:00 CE(S)T YYYY` and in my opinion this is how it should be.

**Comment by cowtowncoder** [comment]
@modrljin one thing that we must be careful about is this:

    deserializing string 1940-04-02 00:00:00 with jackson into cetDatetime 

if there is no timezone -- where is it? -- it is not exactly clear what the timezone to use would be.
Since there MUST be some timezone to use as `java.util.Date` is a concrete physical timestamp and not abstract (or "local" for Java 8 date/times) date/time point.
So different tools can choose different timezone, hence resulting in different actual timestamp/timepoint.

If Jackson does allow such values (IMO it probably should straight fail), it will:

* Not use anything from `@JsonFormat`: timezone is solely for purpose of serializing date/time in that timezone; override what Java 8 / Joda date-time / Calendar might use (that is, just transform from timezone other wise used), or define timezone for serialization in case of `java.util.Date`.
* "Default" timezone from `DeserializationContext` is not used either, I think; rather, UTC is used if nothing is specified in input String

Put another way, value would be handled same as if it was

    1940-04-02T00:00:00Z

**Comment by yawkat** [comment]
@modrljin you're right, it's not actually CEST, it just says so in the Date.toString (I did say java.util.Date timezone handling is a mess :) ). The date is actually in UTC+03:00, which was the daylight savings time here in Berlin during that period. So it's extra confusing.

**Comment by modrljin** [comment]
@cowtowncoder you ask where timezone is in string 1940-04-02 00:00:00, it is nowhere but field has annotation which says 'hay, I want this Date to be treated as CET when serializing a d deserializing'. I will repeat, it really works for all the dates after 1983 and a lot of them before that year, just some years have a problem. Next week I'll check deserialization into other timezones with the same logic string without specifying zone. As far as I have seen the same string when deserialized into UTC gives another and correct Date what means that timezone parameter in annotation does matter and is considered when deserializing, otherwise the same string would give the same Date no matter whi h timezone in annotation is specified, and this is not the case. For dates where deserialization into CET is correct difference to UTC is 1 hour, but for dates where it deserializes into CET wrong difference to UTC is 2 hours. I am talking here about difference between timestamps.

**Comment by cowtowncoder** [comment]
@modrljin First of all, thank you for digging into this and helping troubleshoot the issue. Java date/time handling is a known scourge of developers everywhere... quagmire that keeps on giving.
So anything we can do to lessen the pain is good. But painful task. :)

So. As I have mentioned, the timezone annotation is not intended to be used as "default" timezone on deserialization. It should have no effect on actual value as decoded.
If it does, that in itself would be a bug.
Annotation `timezone` is not intended as part of input data, to augment it, but only as changing output representation (that is, same date/time value but possible written as in timezone different than what Java value has (if any)).
There may be cases for timezone-containing datatypes (java.util.Calendar) where timezone for actual instance would be different, but that should only after decoding has been done using UTC.

So having a unit test, piece of code, would be very useful. And just to be sure (I assume you know this already, apologies for repeating if so), internal timestamp value(s) need to be compared, java.util.Date.toString() avoided (although printing using helper classes the control timezone is fine of course).

I don't know why different dates (after 1983) would behave differently; that does sound interesting.

It does sound like you have found a real problem so above is not trying to argue against there being issues. I just want to ensure we have shared understanding of the situation.

**Comment by modrljin** [comment]
@cowtowncoder thank you for understanding, I'll try to keep digging into the issue with additional tests next week so we all together find the solution or at least some conclusion. Have a nice weekend!

**Comment by modrljin** [comment]
@cowtowncoder @yawkat I think I've found the issue. It seems it is really the problem with java.util.Date because parsing the date with SimpleDateFormat **and TimeZone.getTimeZone("CET")** for all the dates from the periods I have written in my first post is wrong.

```
For example "1940-12-11" is parsed with SimpleDateFormat into:
1. with TimeZone.getTimeZone("CET"): Tue Dec 10 23:00:00 CET 1940 (timestamp: -916970400000)
2. without Timezone: Wed Dec 11 00:00:00 CET 1940 (timestamp: -916966800000)

And Jackson is doing it the same, string "1940-12-11 00:00:00" is deserialized into field with timezone = "CET" into: 
Tue Dec 10 23:00:00 CET 1940
```

I didn't realize it and was getting difference in timestamps because test app was lopping from correct day, e.g. 1.1.1900 which was parsed fine, and then was adding day by day with Calendar function which in fact have created Date: Tue Apr 02 00:00:00 CET 1940 (timestamp: -938826000000), but when string is directly parsed with SimpleDateFormat and TimeZone.getTimeZone("CET") it creates Date: Mon Apr 01 23:00:00 CET 1940 (timestamp: -938829600000), the same as it is the result with Jackson deserialization. OMG, terrible!

So my apoligies about disturbing you, but at least I have lerned something - try to avoid Date object and to add fixes to old apps if necessary :-)

Thank you for your time and best regards,
Marin
